### ISSUE TITLE
Continous Integration tests for Fast AI

### ISSUE THREAD

**Issue by PiotrCzapla** [fastai/fastai#286]
Jeremy 
What would you say for adding some integration tests (https://circleci.com/ or https://travis-ci.org/) so that we can save on manual testing? It will make it easier to introduce small fixes and it will give us some ground to write some proper unit test in the future. 

We could start with simply executing some notebooks and watching for errors, it won't check for GPU errors but at least this is a start and we can use this test on our workstations to execute GPU tests. 

Later we could try to flush out the API in unit test to show the relations between different methods.

Let me know what you  think and if it is ok for you I will the setup a CI on my fork and create a pull request explaining how to plug Travis or Circle CI to your account and repo.

**Comment by hamelsmu** [comment]
I am on board to help with this, if approved.

**Comment by jph00** [comment]
Yeah that would be good. I'd rather start with proper python tests rather than executing notebooks - I think starting the "right way" is important to provide a model for other contributors to follow. E.g this is what we're trying to do for docs right now - show a really good role model for the `fastai.transforms` module, and then let the community follow that for the other modules.

I think that means picking one piece of functionality, or the functionality displayed in one lesson, or picking one class, and trying to show how to test it in a reasonably light-weight but effective way. I do agree that integration tests should be the focus. A few thoughts:

- Tests should check correctness both with and without the GPU (assuming one is available); you can just toggle `fastai.core.USE_GPU` to check both backends
- CI integrations tests should run quickly. That probably means we can't *really* know if an algorithm is fully working (i.e that it actually makes the loss go down when run on real data) since that takes a while for any meaningful dataset. But we could at least do a single batch and check loss isn't NaN
- We could have a longer integration test for each major piece of functionality (e.g. 'fine tune an image classifier CNN') that runs for 30 seconds or so. With a carefully chosen dataset this should be enough to confirm the model is doing something reasonable
- We should include at least *some* unit tests, so folks know what our expectations for unit tests are.

If people submit a PR that changes core functionality like the DataLoader, I'd like to see unit tests with that, since there's a lot of edge cases that are easy to miss and the integration tests won't always identify those issues. E.g if padding isn't done correctly in DataLoader, integration tests might still run, but NLP models might become less accurate.

**Comment by PiotrCzapla** [comment]
I think we are in alignment here. Dummy integration tests are easier (just executing a notebook that is prepared), but I see your point with proper unit-tests, but let's give it a go.  I would pick the transformations as this will benefit the most from the tests and I know the code quite well. 

I will get first steps implemented in a separated branch.

**Comment by PiotrCzapla** [comment]
Here is a first version of travis-ci integeration.
https://travis-ci.org/PiotrCzapla/fastai/builds/363774572

It was still building when I was leaving if this is red or yellow then it means the test are working (the current test fail so the button should be yellow)
[![Build Status](https://travis-ci.org/PiotrCzapla/fastai.svg?branch=master)](https://travis-ci.org/PiotrCzapla/fastai)

**Comment by PiotrCzapla** [comment]
I'm getting segmentation faults probably due to memory limits on travis-ci (4GB per docker container).

**Comment by apiltamang** [comment]
Nice ... despite being a dev. for several years now, I haven't yet used any CI systems. Will be good to track this. Would you mind tagging me in future developments?

Thanks much!

**Comment by PiotrCzapla** [comment]
@Gokkulnath  the gist seems to be broken I see just the JS with plenty of document write :/ (after some fiddling I guess you meant this url: https://gist.github.com/Gokkulnath/31b91f1db29d171471269be07d9177ff)

The point of CI is to check that the dependencies don't collide so having them there may help us in the future, and should not hinder you when you write tests, as the test should be mostly executed locally on your pc /aws. 

to run the tests you simply install `pip install pytest` then run `python -m pytest tests`.  I'm playing with a docker instance to reporduce the crash we can observe at travis-ci and the test are failing so I can't show you how they look like but you will be able to see if they are working or not.

**Comment by PiotrCzapla** [comment]
@jph00  I've seen you talking about this on pytroch issue. I've fixed it by setting matplotlib to use Agg as backend and importing the torch and cv in a right order.

**Comment by PiotrCzapla** [comment]
The build is looking good after some spelling fixes (thx @Gokkulnath) we have first successful build on travis-ci. I've updated the transforms.py pull request #312 so that there are no conflicts, but I still want to get the minimal version of transforms2.py working so that the branch can be merged and the new API can be played within it separate space.

**Comment by PiotrCzapla** [comment]
[![Build Status](https://travis-ci.org/PiotrCzapla/fastai.svg?branch=master)](https://travis-ci.org/PiotrCzapla/fastai)

@jph00  the ci it is working fine, It is bundled with the changes to transform.py but it can be extracted to a new branch if you wish.

**Comment by devforfu** [comment]
I wonder how usually people test deep learning algorithms implementations? I mean, if we're talking about performance metrics and model accuracy verifications, not asserting correct properties values? Quite time consuming process to train model and assert its quality I guess. I mostly trying to do it analytically, but probably someone knows possible alternatives?

**Comment by PiotrCzapla** [comment]
@devforfu I think Jeremy would prefer to talk about this on forum. Here is one article that maybe of interest to you, William implemented some tests for SWA:
https://medium.com/@hortonhearsafoo/adding-a-cutting-edge-deep-learning-training-technique-to-the-fast-ai-library-2cd1dba90a49

The  tests aren't yet  merged to Fast AI because they are bundled with other changes that @jph00  is reluctant to merge, so I will split the pull request to two and once it is merged maybe we start a thread on the forum to organise some ppl to write more tests  to the library. 

I wouldn't be too worried with testing the deep-learning parts yet as we are too early in the process. The Fast.ai does not have any test yet and bugs in regular code like transformations can easily break your code the same way as bugs in deep-learning parts.

**Comment by PiotrCzapla** [comment]
@jph00  once the pull request above is merged you need to get a https://travis-ci.org account for fastai, it is free for opensource it uses your github credentials to pull the right repositories, then you just need to go to your profile and select  fastai for tests then you should be able to see the tests running at this url:
https://travis-ci.org/fastai/fastai

Here is how it should look like:
<img width="1223" alt="screen shot 2018-04-28 at 10 56 33" src="https://user-images.githubusercontent.com/340180/39394548-737cd21c-4ad3-11e8-82dc-ca5d2206aca9.png">
